ml/backend/mlx/mlx.go

```diff
@@ -0,0 +1,1099 @@
+	stridesb := make([]int, len(sb))
+	for i := range sb {
+		stridesb[i] = a2.Stride(i)
+	}
+	C.mlx_matmul(&r,
+		a2.(*Array).a,
+		a.a,
+		ctx.(*Context).stream)
+	return newArray(ctx.(*Context), r)
+}
+
+func (a *Array) MulmatFullPrec(ctx ml.Context, a2 ml.Tensor) ml.Tensor {
+	return a.Mulmat(ctx, a2)
+}
+
+// LayerNorm implements ml.Tensor.
+func (a *Array) LayerNorm(ctx ml.Context, w, b ml.Tensor, eps float32) ml.Tensor {
+	var r C.mlx_array
+	C.mlx_fast_layer_norm(
+		&r,
+		a.a,
+		w.(*Array).a,
+		b.(*Array).a,
+		C.float(eps),
+		ctx.(*Context).stream,
+	)
+	return newArray(ctx.(*Context), r)
+}
+
+// Pad implements ml.Tensor.
+func (a *Array) Pad(ctx ml.Context, shape ...int) ml.Tensor {
+	panic("unimplemented")
+}
+
+// Permute implements ml.Tensor.
+func (a *Array) Permute(ctx ml.Context, shape ...int) ml.Tensor {
+	ndim := min(C.mlx_array_ndim(a.a), C.size_t(len(shape)))
+	var r C.mlx_array
+	sh := make([]C.int, ndim)
+	for i := range ndim {
+		sh[i] = (C.int)(shape[i])
+		if int(sh[i]) >= int(ndim) {
+			slog.Error("Permute error", "tensor", a, "shape", shape)
+			panic("invalid pemute call")
+		}
+	}
+	C.mlx_transpose(
+		&r,
+		a.a,
+		&sh[0],
+		ndim,
+		ctx.(*Context).stream,
+	)
+	return newArray(ctx.(*Context), r)
+}
+
+// RMSNorm implements ml.Tensor.
+func (a *Array) RMSNorm(ctx ml.Context, w ml.Tensor, eps float32) ml.Tensor {
+	var r C.mlx_array
+	C.mlx_fast_rms_norm(
+		&r,
+		a.a,
+		w.(*Array).a,
+		C.float(eps),
+		ctx.(*Context).stream,
+	)
+	return newArray(ctx.(*Context), r)
+}
+
+// Reshape implements ml.Tensor.
+func (a *Array) Reshape(ctx ml.Context, shape ...int) ml.Tensor {
+	cshape := make([]C.int, len(shape))
+	for i, dim := range shape {
+		cshape[i] = C.int(dim)
+	}
+	var r C.mlx_array
+	C.mlx_reshape(&r, a.a, &cshape[0], C.size_t(len(cshape)), ctx.(*Context).stream)
+	return newArray(ctx.(*Context), r)
+}
+
+/* MLX breadcrumb for Fast RoPE
+a (array) – Input array.
+dims (int) – The feature dimensions to be rotated. If the input feature is larger than dims then the rest is left unchanged.
+traditional (bool) – If set to True choose the traditional implementation which rotates consecutive dimensions.
+base (float, optional) – The base used to compute angular frequency for each dimension in the positional encodings. Exactly one of base and freqs must be None.
+scale (float) – The scale used to scale the positions.
+offset (int or array) – The position offset to start at.
+freqs (array, optional) – Optional frequencies to use with RoPE. If set, the base parameter must be None. Default: None.
+*/
+
+// Rope implements ml.Tensor.
+func (a *Array) RoPE(
+	ctx ml.Context,
+	positionIDs ml.Tensor, // Unused in MLX
+	ropeFactors ml.Tensor, // Unused in MLX
+	freqs ml.Tensor,
+	dim uint32,
+	ropeType uint32,
+	base float32,
+	scale float32,
+) ml.Tensor {
+	a = a.Reshape(ctx, append([]int{1}, a.Shape()...)...).Permute(ctx, 0, 2, 1, 3).(*Array)
+	// TODO figure out how to get offset wired up
+	offset := 0
```

Comment:
Not using positionIDs is probably also part of the issue where things fall apart after a few forward passes. The prompt will get correctly RoPEd since it starts at offset 0 and is all in one batch. However, every token after that will be at position 0 as well.

From what I saw from quickly looking before, it looks like `offset` is only a scalar in the C interface of MLX but it can be a vector in the Python version, same as we have here:
https://ml-explore.github.io/mlx/build/html/python/_autosummary/mlx.core.fast.rope.html